aws-analytics-pinpoint/src/main/java/com/amplifyframework/analytics/pinpoint/AmazonPinpointAnalyticsPlugin.java

```diff
@@ -79,8 +93,71 @@ public void enable() {
+    }
+
+    /**
+     * Add user specific data from {@link AnalyticsUserProfile} to the endpoint profile.
+     * @param endpointProfile endpoint profile.
+     * @param userProfile user specific data to be added to the endpoint.
+     */
+    private void addUserProfileToEndpoint(EndpointProfile endpointProfile,
+                                          AnalyticsUserProfile userProfile) {
+        endpointProfile.addAttribute("email", Arrays.asList(userProfile.getEmail()));
+        endpointProfile.addAttribute("name", Arrays.asList(userProfile.getName()));
+        endpointProfile.addAttribute("plan", Arrays.asList(userProfile.getPlan()));
+        // Add location
+        addLocation(endpointProfile.getLocation(), userProfile.getLocation());
+        // Add custom properties
+        addCustomProperties(endpointProfile, userProfile.getCustomProperties());
+    }
+
+    /**
+     * Add custom user properties to the endpoint profile.
+     * @param endpointProfile endpoint profile.
+     * @param customProperties custom user properties to be added to the endpoint profile.
+     */
+    private void addCustomProperties(EndpointProfile endpointProfile,
+                                     Properties customProperties) {
+        if (customProperties != null) {
+            for (Map.Entry<String, Property<?>> entry : customProperties.get().entrySet()) {
+                if (entry.getValue() instanceof StringProperty) {
+                    endpointProfile.addAttribute(entry.getKey(),
+                            Arrays.asList(((StringProperty) entry.getValue()).getValue()));
+                } else if (entry.getValue() instanceof DoubleProperty) {
+                    endpointProfile.addMetric(entry.getKey(),
+                            ((DoubleProperty) entry.getValue()).getValue());
+                } else {
+                    Amplify.Hub.publish(HubChannel.ANALYTICS, HubEvent.create("Analytics.identifyUser",
+                            "Invalid property type detected. AmazonPinpointAnalyticsPlugin supports" +
+                                    " only StringProperty or DoubleProperty. Refer to the documentation for details."));
+                }
+            }
+        }
+    }
+
+    /**
+     * Add location details to the endpoint profile location.
+     * @param endpointProfileLocation endpoint location.
+     * @param location location details.
+     */
+    private void addLocation(EndpointProfileLocation endpointProfileLocation,
+                             AnalyticsUserProfile.Location location) {
+        endpointProfileLocation.setLatitude(location.getLatitude());
+        endpointProfileLocation.setLongitude(location.getLongitude());
+        endpointProfileLocation.setPostalCode(location.getPostalCode());
+        endpointProfileLocation.setCity(location.getCity());
+        endpointProfileLocation.setRegion(location.getRegion());
+        endpointProfileLocation.setCountry(location.getCountry());
```

Comment:
Looks like you want a builder. :-D If some of these fields are required, then consider a step builder. (That way, you could check if required fields are present at build time, not just runtime.)

Reply:
I agree `EndpointProfileLocation` can benefit from Builder because of so many optional params. We discussed not to do it since EndpointProfileLocation is a part of pinpoint sdk.